### PR TITLE
feat: Support lora in dtensor grpo workflow by merging weight

### PR DESCRIPTION
**Pull request by RayenTian**
## Summary
- Merge LoRA adapter weights into base linear weights when exporting dtensor state and skip standalone LoRA adapter tensors.
- Add LoRA configuration defaults to `grpo_math_1B.yaml` and introduce a Qwen3-8B LoRA recipe.
- Expand LoRA coverage in functional and unit tests (vLLM generation + GRPO LoRA suites).

## Issues
closes https://github.com/NVIDIA-NeMo/RL/issues/1597

## Changes
- `nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py`
  - Merge LoRA weights into base weights during state export.
  - Skip `lora_A/lora_B` tensors and release temporary tensors to reduce memory.
- `examples/configs/grpo_math_1B.yaml`
  - Add LoRA config section with defaults/documentation.
- `examples/configs/recipes/llm/grpo-qwen3-8B-base-1n8g-fsdp2-lora.yaml`
  - New LoRA recipe for Qwen3-8B.
- `tests/functional/*`
  - Add GRPO LoRA functional tests (sync/async/non-colocated) and include in nightly.
- `tests/unit/models/generation/test_vllm_generation.py`
  - Add LoRA config coverage and parameters in vLLM tests.

# The compare of two implementation approach: Split Weight and Merge Weight
### PRs:

- https://github.com/NVIDIA-NeMo/RL/pull/1748
- https://github.com/NVIDIA-NeMo/RL/pull/1751
- https://github.com/NVIDIA-NeMo/RL/pull/1752

### Design

Enable Lora for both vllm and dtensor backend. Update base model weight and lora weight at the first step, and only refit lora weights in the following step.

Both inference and train compute with  $xW^T + scale\times x (BA)^T$

- The code is more complex. Additional processing is required for the refit process.
- ❗ VLLM and AutoModel handle LoRA addition to the `lm_head` differently for various models. A patch is needed for VLLM to skip the `lm_head`. (refer to https://nvidia.slack.com/archives/C0A1J3A7G85/p1768803141238469)
- ❗ Performance degrades because VLLM computes much slower with LoRA enabled.

### Results:

- Time

  | Qwen3-0.6B<br />lora_dim(128) <br />SeqLen(4096)<br />step (33) | Total Time(s)/Step | Generation Time(s)/Step | Train Time(s)/Step | Refit Time(s)/Step |
  | ------------------------------------------------------------ | ------------------ | ----------------------- | ------------------ | ------------------ |
  | W/O Lora                                                     | 55                 | 37                      | 10.2               | 0.4                |
  | Lora Split Weight                                            | 67                 | 48                      | 9.9                | 0.5                |

  - Enabling LoRA in vLLM leads to a significant performance drop.
  - Refit Latency: Even with our current approach of only updating LoRA weights, we still see a slight increase in refit time.

### PR:

- https://github.com/NVIDIA-NeMo/RL/pull/1797

### Design

Only enable lora in dtensor backend and VLLM is **agnostic** to LoRA. Merge base model weight and lora weight before refit each time.

Dtensor compute lora via  $xW^T + scale\times x (BA)^T$

Vllm compute lora via   $x W_{vllm}^T= x (W^T + scale\times(BA)^T)$

- 👍The code is simpler; no extra consideration is needed for `lm_head` handling across different frameworks.
- 👍Superior performance.
- ❗Extra **training-inference discrepancy** is introduced because the calculations of $xW^T + scale\times x (BA)^T$ and $x W_{vllm}^T= x (W^T + scale\times(BA)^T)$ are not **bitwise identical** across hardware.
  - Post several dozen training steps, the `train/gen_kl_error` exhibits a slight mismatch. Notably, this divergence does not escalate without bound but rather stabilizes at a certain level.
  - Can observe this result on both qwen and llama.
  - Our findings confirm that identical data and the NeMo RL DTensor backend do not guarantee identical logprobs. Specifically, a non-zero KL divergence is observed between the merged approach (base + LoRA weights combined upfront) and the additive approach (base and LoRA outputs computed independently and summed).
  - However, this `gen_kl_error` can be controlled to within 0.001, so we believe this is acceptable.

### Result

| Qwen3-0.6B <br />lora_dim(128)<br />SeqLen(4096)<br />step (33) | Total Time(s)/Step | Generation Time(s)/Step | Train Time(s)/Step | Refit Time(s)/Step |
| ------------------------------------------------------------ | ------------------ | ----------------------- | ------------------ | ------------------ |
| W/O Lora                                                     | 55                 | 37                      | 10.2               | 0.4                |
| Lora Split Weight                                            | 67                 | 48                      | 9.9                | 0.5                |
| Lora Merge Weight                                            | 56.9               | 36.58                   | 10.57              | 0.549              |

| llama3.1-8B <br />lora_dim(128) <br />SeqLen(1024)<br />step (33) | Total Time(s)/Step | Generation Time(s)/Step | Train Time(s)/Step | Refit Time(s)/Step |
| ------------------------------------------------------------ | ------------------ | ----------------------- | ------------------ | ------------------ |
| W/O Lora                                                     | 28.82              | 10.4                    | 8.55               | 0.54               |
| Lora Merge Weight                                            | 30.98              | 10.9                    | 7.6                | 0.74               |

## Convergence Testing
image

## Summary by CodeRabbit

* **New Features**
  * Added LoRA (Low-Rank Adaptation) support for policy training with configurable rank, scaling, and dropout options
  * New Qwen3 8B model configuration with LoRA enabled for GRPO experiments
  * Added validation to prevent incompatible LoRA configurations

* **Tests**
  * Added comprehensive functional and unit tests for LoRA scenarios including async and non-colocated training configurations

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/models/policy/workers/patches.py [hunk 1]
```diff
@@ -126,3 +127,25 @@ def apply_torch_aten_alias_tensor_patch():
         )
     except Exception as e:
         print(f"Error applying torch.ops.aten.alias.default patch: {e}")
+
+
+def patched_lora_linear_forward(self, x):
```
Comment (RayenTian) [inline nemo_rl/models/policy/workers/patches.py]: this patch cannot guarantee that the computational logic is exactly identical to the [original](https://github.com/NVIDIA-NeMo/Automodel/blob/f85e4203155349385c876f3289af7f799c47b494/nemo_automodel/components/_peft/lora.py#L189).

File: nemo_rl/models/automodel/setup.py [hunk 2]
```diff
@@ -479,6 +479,22 @@ def setup_model_and_optimizer(
...
         model = distributed_manager.parallelize(model)
 
+    if lora_enabled:
+        # If the model uses quantized weights, we want to use orig_linear's forward
+        try:
+            import bitsandbytes as bnb
+        except Exception:  # pragma: no cover - optional dependency
+            bnb = None
+
+        if bnb is not None:
```
Comment (yuki-97) [inline nemo_rl/models/automodel/setup.py]: do we always have this dependency installed? if so, feels we could remove the try except?

Comment (RayenTian) [reply nemo_rl/models/automodel/setup.py]: This is add for patch. Now I removed the patch, so this code doesn't need any more.

**Review by yuki-97 — APPROVED** [review]
